File: internal/radio/remote.go
package radio

import (
	"context"
	"crypto/rand"
	"crypto/subtle"
	_ "embed"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"slices"
	"strings"
	"sync"
	"time"
)

//go:embed remote.html
var remoteHTML []byte

const (
	allStationsTag = "All Stations"
	tokenAlphabet  = "abcdefghijkmnpqrstuvwxyz23456789"
	tokenLength    = 6
)

var errNotFound = errors.New("not found")

// Remote is the HTTP server behind Ctrl+P. It lives until the TUI exits.
type Remote struct {
	app   *Application
	token string // empty: no access code is asked for
	port  int
	ip    string
	srv   *http.Server

	mu         sync.Mutex
	lastSearch remoteSearch
}

type remoteSearch struct {
	query    string
	online   bool
	stations []Station
}

type remoteTag struct {
	Name   string `json:"name"`
	Kind   string `json:"kind,omitempty"`
	Online bool   `json:"online,omitempty"`
}

type remoteStation struct {
	Title   string `json:"title"`
	URL     string `json:"url"`
	Meta    string `json:"meta,omitempty"`
	Playing bool   `json:"playing"`
}

type remotePlayer struct {
	Status  string `json:"status"`
	Station string `json:"station"`
	Song    string `json:"song"`
	URL     string `json:"url"`
	Volume  int    `json:"volume"`
	Bitrate int    `json:"bitrate"`
}

type remoteShuffle struct {
	Active    bool `json:"active"`
	Remaining int  `json:"remaining"`
	Interval  int  `json:"interval"`
}

type remoteState struct {
	Version  string          `json:"version"`
	Page     string          `json:"page"`
	Tag      string          `json:"tag"`
	Tags     []remoteTag     `json:"tags"`
	Stations []remoteStation `json:"stations"`
	Player   remotePlayer    `json:"player"`
	Shuffle  remoteShuffle   `json:"shuffle"`
}

// startRemote is idempotent: a running server is returned as is.
func (a *Application) startRemote() (*Remote, error) {
	if a.remote != nil {
		return a.remote, nil
	}

	ln, err := net.Listen("tcp", fmt.Sprintf(":%d", a.remotePort))
	if err != nil {
		log.Printf("remote: port %d unavailable (%v), picking a free one", a.remotePort, err)
		if ln, err = net.Listen("tcp", ":0"); err != nil {
			return nil, err
		}
	}

	token := newToken(tokenLength)
	if a.remoteKey != nil {
		token = strings.TrimSpace(*a.remoteKey)
	}

	r := &Remote{
		app:   a,
		token: token,
		port:  ln.Addr().(*net.TCPAddr).Port,
		ip:    lanIP(),
	}

	r.srv = &http.Server{Handler: r.handler(), ReadHeaderTimeout: 10 * time.Second}
	go func() {
		if err := r.srv.Serve(ln); err != nil && !errors.Is(err, http.ErrServerClosed) {
			log.Printf("remote: server stopped: %v", err)
		}
	}()

	log.Printf("remote: listening on %s", r.URL())
	a.remote = r
	return r, nil
}

func (a *Application) stopRemote() {
	if a.remote == nil {
		return
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	if err := a.remote.srv.Shutdown(ctx); err != nil {
		a.remote.srv.Close()
	}
	log.Println("remote: server stopped")
	a.remote = nil
}

// Address is what a user types into the phone browser; URL is what the QR
// code carries, key included.
func (r *Remote) Address() string {
	return "http://" + net.JoinHostPort(r.ip, fmt.Sprint(r.port))
}

func (r *Remote) URL() string {
	if r.token == "" {
		return r.Address()
	}
	return r.Address() + "/?k=" + url.QueryEscape(r.token)
}

func (r *Remote) Key() string {
	return r.token
}

func (r *Remote) handler() http.Handler {
	a := r.app
	mux := http.NewServeMux()
	mux.HandleFunc("GET /{$}", r.index)
	mux.HandleFunc("GET /api/state", r.auth(r.state))
	mux.HandleFunc("GET /api/search", r.auth(r.search))
	mux.HandleFunc("POST /api/tags", r.auth(r.action(a.remoteShowTags)))
	mux.HandleFunc("POST /api/tag", r.auth(r.action(a.remoteOpenTag)))
	mux.HandleFunc("POST /api/play", r.auth(r.action(a.remotePlay)))
	mux.HandleFunc("POST /api/stop", r.auth(r.action(a.remoteStop)))
	mux.HandleFunc("POST /api/random", r.auth(r.action(a.remoteRandom)))
	mux.HandleFunc("POST /api/volume", r.auth(r.action(a.remoteVolume)))
	mux.HandleFunc("POST /api/shuffle", r.auth(r.action(a.remoteShuffle)))
	mux.HandleFunc("POST /api/search/select", r.auth(r.action(r.selectSearch)))
	mux.HandleFunc("/favicon.ico", func(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(http.StatusNoContent) })
	return mux
}

type actionRequest struct {
	Tag     string `json:"tag"`
	URL     string `json:"url"`
	Volume  *int   `json:"volume"`
	Minutes int    `json:"minutes"`
	Query   string `json:"query"`
	Online  bool   `json:"online"`
}

// auth accepts the key as the "k" query parameter (QR link) or the X-Key
// header (page script). Case is ignored: phone keyboards capitalize.
func (r *Remote) auth(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		key := req.URL.Query().Get("k")
		if key == "" {
			key = req.Header.Get("X-Key")
		}
		key = strings.ToLower(strings.TrimSpace(key))
		if r.token != "" && subtle.ConstantTimeCompare([]byte(key), []byte(strings.ToLower(r.token))) != 1 {
			http.Error(w, "Forbidden: wrong access code. The code is shown by goradion when you press Ctrl+P.", http.StatusForbidden)
			return
		}
		w.Header().Set("Cache-Control", "no-store")
		next(w, req)
	}
}

func (r *Remote) index(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.Write(remoteHTML)
}

func (r *Remote) state(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, r.app.remoteState())
}

// action runs a state-changing request and replies with the resulting state.
func (r *Remote) action(run func(q actionRequest) error) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		var q actionRequest
		if req.ContentLength > 0 {
			if err := json.NewDecoder(req.Body).Decode(&q); err != nil {
				http.Error(w, "bad request: "+err.Error(), http.StatusBadRequest)
				return
			}
		}
		if err := run(q); err != nil {
			status := http.StatusBadRequest
			if errors.Is(err, errNotFound) {
				status = http.StatusNotFound
			}
			http.Error(w, err.Error(), status)
			return
		}
		writeJSON(w, r.app.remoteState())
	}
}

func (r *Remote) search(w http.ResponseWriter, req *http.Request) {
	query := strings.TrimSpace(req.URL.Query().Get("q"))
	online := req.URL.Query().Get("online") == "1"

	var results []remoteStation
	var stations []Station

	if query != "" && online {
		found, err := SearchRadioBrowser(query)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadGateway)
			return
		}
		for _, f := range found {
			stations = append(stations, f.station)
			results = append(results, remoteStation{Title: f.station.title, URL: f.station.url, Meta: onlineMeta(f)})
		}
	} else if query != "" {
		stations = r.app.filterStations(query)
		for _, s := range stations {
			results = append(results, remoteStation{Title: s.title, URL: s.url})
		}
	}

	r.mu.Lock()
	r.lastSearch = remoteSearch{query: query, online: online, stations: stations}
	r.mu.Unlock()

	if results == nil {
		results = []remoteStation{}
	}
	writeJSON(w, map[string]any{"query": query, "online": online, "results": results})
}

// selectSearch does what picking a result in the TUI search modal does: the
// results become the current list and the chosen station starts playing.
func (r *Remote) selectSearch(q actionRequest) error {
	r.mu.Lock()
	last := r.lastSearch
	r.mu.Unlock()

	if last.query == "" || last.query != q.Query || last.online != q.Online {
		return errors.New("search results are stale, search again")
	}

	for _, s := range last.stations {
		if s.url == q.URL {
			r.app.app.QueueUpdateDraw(func() {
				r.app.openSearchInMain(last.query, last.stations, last.online)
				r.app.stationsList.SetCurrentItem(r.app.findStationIndex(s.url, last.stations))
			})
			r.app.togglePlayManual(s)
			return nil
		}
	}
	return fmt.Errorf("station %w in search results", errNotFound)
}

func writeJSON(w http.ResponseWriter, v any) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v)
}

// remoteState snapshots what the TUI shows. UI fields are read on the UI
// goroutine; the player lock is taken afterwards, never inside the queued
// function: the player blocks on the Info channel while holding its lock, and
// the channel's consumer needs the UI loop to draw.
func (a *Application) remoteState() remoteState {
	st := remoteState{Version: VersionString(), Page: "tags"}

	var stations []Station
	a.app.QueueUpdate(func() {
		st.Tag = a.tag
		st.Tags = a.remoteTags()
		if slices.Contains(a.pages.GetPageNames(true), a.pageNames[Main]) {
			st.Page = "stations"
			stations = a.getStationsFromCurrentView()
		}
		st.Shuffle.Interval = int(a.shuffleInterval.Minutes())
		if a.timedRandomActive {
			st.Shuffle.Active = true
			st.Shuffle.Remaining = int(max(a.shuffleInterval-time.Since(a.shuffleIterationStartAt), 0).Seconds())
		}
	})

	a.player.Lock()
	inf := *a.player.info
	a.player.Unlock()

	st.Player = remotePlayer{
		Status:  inf.Status,
		Station: stripPlayCount(inf.Station),
		Song:    inf.Song,
		URL:     inf.Url,
		Volume:  inf.Volume,
		Bitrate: inf.Bitrate,
	}

	st.Stations = make([]remoteStation, 0, len(stations))
	for _, s := range stations {
		st.Stations = append(st.Stations, remoteStation{
			Title:   stripPlayCount(s.title),
			URL:     s.url,
			Playing: s.url != "" && s.url == inf.Url,
		})
	}
	return st
}

func (a *Application) remoteTags() []remoteTag {
	out := make([]remoteTag, 0)
	if a.favorites.hasFavorites() {
		out = append(out, remoteTag{Name: favoritesTag, Kind: "favorites"})
	}
	out = append(out, remoteTag{Name: allStationsTag, Kind: "all"})
	for _, t := range tags(a.stations) {
		out = append(out, remoteTag{Name: t})
	}
	if a.lastSearchTag != "" {
		out = append(out, remoteTag{Name: a.lastSearchTag, Kind: "search", Online: a.lastOnlineStations != nil})
	}
	return out
}

func (a *Application) remoteShowTags(actionRequest) error {
	a.app.QueueUpdateDraw(func() {
		a.tag = ""
		a.show(Tags)
	})
	return nil
}

func (a *Application) remoteOpenTag(q actionRequest) error {
	if q.Tag == favoritesTag && !a.favorites.hasFavorites() {
		return fmt.Errorf("tag %q %w", q.Tag, errNotFound)
	}
	var err error
	a.app.QueueUpdateDraw(func() {
		if !a.openTag(q.Tag) {
			err = fmt.Errorf("tag %q %w", q.Tag, errNotFound)
		}
	})
	return err
}

// remotePlay toggles a station of the list currently shown, like Enter does.
func (a *Application) remotePlay(q actionRequest) error {
	var station Station
	var found bool
	a.app.QueueUpdateDraw(func() {
		stations := a.getStationsFromCurrentView()
		for i, s := range stations {
			if s.url == q.URL {
				station, found = s, true
				a.stationsList.SetCurrentItem(i + a.calculateStationListOffset())
				return
			}
		}
	})
	if !found {
		return fmt.Errorf("station %w in the current list", errNotFound)
	}
	a.togglePlayManual(station)
	return nil
}

func (a *Application) remoteStop(actionRequest) error {
	a.player.Lock()
	current := Station{title: a.player.info.Station, url: a.player.info.Url}
	a.player.Unlock()

	if current.url == "" {
		return nil
	}
	a.togglePlayManual(current)
	return nil
}

func (a *Application) remoteRandom(actionRequest) error {
	var station Station
	var ok bool
	a.app.QueueUpdateDraw(func() {
		stations := a.getStationsFromCurrentView()
		if len(stations) == 0 {
			return
		}
		r := a.randomIndex(stations)
		a.stationsList.SetCurrentItem(r + a.calculateStationListOffset())
		station, ok = stations[r], true
	})
	if !ok {
		return errors.New("no stations to pick from")
	}
	a.togglePlayManual(station)
	return nil
}

func (a *Application) remoteVolume(q actionRequest) error {
	if q.Volume == nil {
		return errors.New("volume required")
	}
	a.player.SetVolume(*q.Volume)
	return nil
}

func (a *Application) remoteShuffle(q actionRequest) error {
	if q.Minutes != 0 {
		if q.Minutes < 1 || q.Minutes > 9 {
			return errors.New("minutes must be 1-9")
		}
		a.setShuffleInterval(q.Minutes)
		return nil
	}
	a.toggleTimedRandom()
	return nil
}

func newToken(n int) string {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		panic(err)
	}
	for i := range b {
		b[i] = tokenAlphabet[int(b[i])%len(tokenAlphabet)]
	}
	return string(b)
}

// lanIP is the default-route IPv4 address, else the first one of an interface
// that is up, else loopback.
func lanIP() string {
	usable := func(ip net.IP) bool {
		ip = ip.To4()
		return ip != nil && !ip.IsLoopback() && !ip.IsLinkLocalUnicast()
	}

	// Connecting a UDP socket only resolves the route; nothing is sent.
	if conn, err := net.Dial("udp4", "8.8.8.8:80"); err == nil {
		ip := conn.LocalAddr().(*net.UDPAddr).IP
		conn.Close()
		if usable(ip) {
			return ip.String()
		}
	}

	if ifaces, err := net.Interfaces(); err == nil {
		for _, iface := range ifaces {
			if iface.Flags&net.FlagUp == 0 || iface.Flags&net.FlagLoopback != 0 {
				continue
			}
			addrs, err := iface.Addrs()
			if err != nil {
				continue
			}
			for _, addr := range addrs {
				if ipn, ok := addr.(*net.IPNet); ok && usable(ipn.IP) {
					return ipn.IP.String()
				}
			}
		}
	}

	return "127.0.0.1"
}

File: rust/src/remote.rs
//! The phone remote's HTTP server. Requests that read or change the app are
//! sent to the UI loop as Calls and answered with the resulting state.

use std::collections::HashMap;
use std::io::Read;
use std::net::UdpSocket;
use std::sync::mpsc::{self, Receiver, Sender};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

use serde::{Deserialize, Serialize};
use tiny_http::{Header, Method, Request, Response};

use crate::log;
use crate::radiobrowser;
use crate::stations::Station;

const PAGE: &str = include_str!("../../internal/radio/remote.html");
const TOKEN_ALPHABET: &[u8] = b"abcdefghijkmnpqrstuvwxyz23456789";
pub const TOKEN_LENGTH: usize = 6;
const MAX_REQUEST_BYTES: u64 = 64 << 10;
const REPLY_TIMEOUT: Duration = Duration::from_secs(10);

#[derive(Debug)]
pub enum Action {
    State,
    Tags,
    Tag,
    Play,
    Bookmark,
    Stop,
    Random,
    Volume,
    Shuffle,
    Sleep,
    /// Picks a result of the last search served, which it carries.
    SelectSearch(LastSearch),
}

#[derive(Deserialize, Default, Debug)]
#[serde(default)]
pub struct ActionRequest {
    pub tag: String,
    /// The tag is the last search.
    pub search: bool,
    pub url: String,
    pub volume: Option<i32>,
    pub minutes: i64,
    pub query: String,
    pub online: bool,
}

pub struct Call {
    pub action: Action,
    pub req: ActionRequest,
    pub reply: Sender<Result<State, ApiError>>,
}

#[derive(Debug)]
pub struct ApiError {
    pub status: u16,
    pub message: String,
}

impl ApiError {
    pub fn bad(message: impl Into<String>) -> Self {
        ApiError { status: 400, message: message.into() }
    }

    pub fn not_found(message: impl Into<String>) -> Self {
        ApiError { status: 404, message: message.into() }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
pub struct Tag {
    pub name: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub kind: String,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub online: bool,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
pub struct StationView {
    pub title: String,
    pub url: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub meta: String,
    #[serde(default)]
    pub playing: bool,
    #[serde(default)]
    pub bookmarked: bool,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
pub struct PlayerView {
    pub status: String,
    pub station: String,
    pub song: String,
    pub url: String,
    pub volume: i32,
    pub bitrate: u32,
    pub bookmarked: bool,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
pub struct ShuffleView {
    pub active: bool,
    pub remaining: u64,
    pub interval: u64,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
pub struct SleepView {
    pub active: bool,
    pub remaining: u64,
}

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
pub struct State {
    pub version: String,
    pub page: String,
    pub tag: String,
    pub tags: Vec<Tag>,
    pub stations: Vec<StationView>,
    pub player: PlayerView,
    pub shuffle: ShuffleView,
    pub sleep: SleepView,
}

#[derive(Clone, Debug, Default)]
pub struct LastSearch {
    pub query: String,
    pub online: bool,
    pub stations: Vec<Station>,
}

struct Shared {
    /// Empty: no access code is asked for.
    token: String,
    stations: Vec<Station>,
    calls: Mutex<Sender<Call>>,
    last_search: Mutex<LastSearch>,
}

/// Lives until dropped, which stops it.
pub struct Server {
    http: Arc<tiny_http::Server>,
    shared: Arc<Shared>,
    pub port: u16,
    pub ip: String,
    thread: Option<thread::JoinHandle<()>>,
}

impl Server {
    /// Listens on all interfaces at port, or a free port when it is taken. A
    /// None key makes a random code.
    pub fn start(port: u16, key: Option<&str>, stations: Vec<Station>) -> std::io::Result<(Server, Receiver<Call>)> {
        let http = tiny_http::Server::http(("0.0.0.0", port))
            .or_else(|e| {
                log!("remote: port {port} unavailable ({e}), picking a free one");
                tiny_http::Server::http(("0.0.0.0", 0))
            })
            .map_err(std::io::Error::other)?;
        let port = http.server_addr().to_ip().map_or(port, |a| a.port());
        let token = key.map_or_else(|| new_token(TOKEN_LENGTH), |k| k.trim().to_string());
        let (tx, rx) = mpsc::channel();
        let shared =
            Arc::new(Shared { token, stations, calls: Mutex::new(tx), last_search: Mutex::new(LastSearch::default()) });
        let http = Arc::new(http);
        let (serve_http, serve_shared) = (http.clone(), shared.clone());
        let thread = thread::Builder::new().name("remote".into()).spawn(move || {
            for req in serve_http.incoming_requests() {
                let shared = serve_shared.clone();
                let _ = thread::Builder::new().name("remote-request".into()).spawn(move || handle(req, &shared));
            }
        })?;
        let server = Server { http, shared, port, ip: lan_ip(), thread: Some(thread) };
        log!("remote: listening on {}", server.url());
        Ok((server, rx))
    }

    /// What a user types into the phone browser.
    pub fn address(&self) -> String {
        format!("http://{}:{}", self.ip, self.port)
    }

    /// What the QR code carries, key included.
    pub fn url(&self) -> String {
        if self.shared.token.is_empty() {
            return self.address();
        }
        let key: String = url::form_urlencoded::byte_serialize(self.shared.token.as_bytes()).collect();
        format!("{}/?k={key}", self.address())
    }

    pub fn key(&self) -> &str {
        &self.shared.token
    }
}

impl Drop for Server {
    fn drop(&mut self) {
        self.http.unblock();
        if let Some(t) = self.thread.take() {
            let _ = t.join();
        }
        log!("remote: server stopped");
    }
}

fn handle(req: Request, shared: &Shared) {
    let method = req.method().clone();
    let url = url::Url::parse(&format!("http://localhost{}", req.url())).ok();
    let path = url.as_ref().map_or("/".to_string(), |u| u.path().to_string());
    let query: HashMap<String, String> = url.map(|u| u.query_pairs().into_owned().collect()).unwrap_or_default();

    let response = match (&method, path.as_str()) {
        (Method::Get, "/") => html(PAGE),
        (_, "/favicon.ico") => Response::from_string("").with_status_code(204),
        (_, p) if p.starts_with("/api/") => {
            if !authorized(shared, &req, &query) {
                text(403, "Forbidden: wrong access code. The code is shown by goradion when you press Ctrl+P.")
            } else {
                api(req, shared, method, &path, &query);
                return;
            }
        }
        _ => text(404, "404 page not found"),
    };
    let _ = req.respond(response);
}

/// The key comes as the "k" query parameter (QR link) or the X-Key header
/// (page script). Case is ignored: phone keyboards capitalize.
fn authorized(shared: &Shared, req: &Request, query: &HashMap<String, String>) -> bool {
    if shared.token.is_empty() {
        return true;
    }
    let header = req.headers().iter().find(|h| h.field.equiv("X-Key")).map(|h| h.value.as_str().to_string());
    let key = query.get("k").filter(|k| !k.is_empty()).cloned().or(header).unwrap_or_default();
    constant_time_eq(key.trim().to_lowercase().as_bytes(), shared.token.to_lowercase().as_bytes())
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    a.len() == b.len() && a.iter().zip(b).fold(0u8, |acc, (x, y)| acc | (x ^ y)) == 0
}

fn api(mut req: Request, shared: &Shared, method: Method, path: &str, query: &HashMap<String, String>) {
    let action = match (method, path) {
        (Method::Get, "/api/state") => Action::State,
        (Method::Get, "/api/search") => {
            let _ = req.respond(search(shared, query));
            return;
        }
        (Method::Post, "/api/tags") => Action::Tags,
        (Method::Post, "/api/tag") => Action::Tag,
        (Method::Post, "/api/play") => Action::Play,
        (Method::Post, "/api/bookmark") => Action::Bookmark,
        (Method::Post, "/api/stop") => Action::Stop,
        (Method::Post, "/api/random") => Action::Random,
        (Method::Post, "/api/volume") => Action::Volume,
        (Method::Post, "/api/shuffle") => Action::Shuffle,
        (Method::Post, "/api/sleep") => Action::Sleep,
        (Method::Post, "/api/search/select") => Action::SelectSearch(shared.last_search.lock().unwrap().clone()),
        _ => {
            let _ = req.respond(text(404, "404 page not found"));
            return;
        }
    };
    let mut body = String::new();
    if let Err(e) = req.as_reader().take(MAX_REQUEST_BYTES).read_to_string(&mut body) {
        let _ = req.respond(text(400, &format!("bad request: {e}")));
        return;
    }
    let parsed = if body.trim().is_empty() { Ok(ActionRequest::default()) } else { serde_json::from_str(&body) };
    let action_req = match parsed {
        Ok(r) => r,
        Err(e) => {
            let _ = req.respond(text(400, &format!("bad request: {e}")));
            return;
        }
    };

    let (tx, rx) = mpsc::channel();
    let sent = shared.calls.lock().unwrap().send(Call { action, req: action_req, reply: tx }).is_ok();
    let response = match rx.recv_timeout(REPLY_TIMEOUT) {
        Ok(Ok(state)) if sent => json(&state),
        Ok(Err(e)) => text(e.status, &e.message),
        _ => text(503, "goradion is exiting"),
    };
    let _ = req.respond(response);
}

fn search(shared: &Shared, query: &HashMap<String, String>) -> Response<std::io::Cursor<Vec<u8>>> {
    let q = query.get("q").map_or("", |s| s.trim()).to_string();
    let online = query.get("online").is_some_and(|v| v == "1");
    let (stations, results): (Vec<Station>, Vec<StationView>) = if !q.is_empty() && online {
        match radiobrowser::search(&q) {
            Ok(found) => found
                .into_iter()
                .map(|f| {
                    let meta = crate::ui::online_meta(&f);
                    let view = StationView {
                        title: f.station.title.clone(),
                        url: f.station.url.clone(),
                        meta,
                        ..StationView::default()
                    };
                    (f.station, view)
                })
                .unzip(),
            Err(e) => return text(502, &e),
        }
    } else {
        crate::ui::search_stations(&shared.stations, &q)
            .into_iter()
            .map(|s| {
                let view = StationView { title: s.title.clone(), url: s.url.clone(), ..StationView::default() };
                (s, view)
            })
            .unzip()
    };
    *shared.last_search.lock().unwrap() = LastSearch { query: q.clone(), online, stations };
    json(&serde_json::json!({ "query": q, "online": online, "results": results }))
}

fn header(name: &str, value: &str) -> Header {
    Header::from_bytes(name.as_bytes(), value.as_bytes()).expect("valid header")
}

fn html(body: &str) -> Response<std::io::Cursor<Vec<u8>>> {
    Response::from_string(body).with_header(header("Content-Type", "text/html; charset=utf-8"))
}

fn json(v: &impl Serialize) -> Response<std::io::Cursor<Vec<u8>>> {
    let body = serde_json::to_string(v).unwrap_or_else(|_| "{}".into());
    Response::from_string(body)
        .with_header(header("Content-Type", "application/json"))
        .with_header(header("Cache-Control", "no-store"))
}

fn text(status: u16, message: &str) -> Response<std::io::Cursor<Vec<u8>>> {
    Response::from_string(format!("{message}\n"))
        .with_status_code(status)
        .with_header(header("Content-Type", "text/plain; charset=utf-8"))
}

pub fn new_token(n: usize) -> String {
    use rand::RngExt;
    let mut rng = rand::rng();
    (0..n).map(|_| TOKEN_ALPHABET[rng.random_range(0..TOKEN_ALPHABET.len())] as char).collect()
}

/// The default-route IPv4 address, else loopback. Connecting a UDP socket
/// only resolves the route; nothing is sent.
pub fn lan_ip() -> String {
    UdpSocket::bind("0.0.0.0:0")
        .and_then(|s| s.connect("8.8.8.8:80").map(|_| s))
        .and_then(|s| s.local_addr())
        .ok()
        .map(|a| a.ip())
        .filter(|ip| !ip.is_loopback() && !ip.is_unspecified())
        .map_or_else(|| "127.0.0.1".to_string(), |ip| ip.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tokens() {
        let mut seen = std::collections::HashSet::new();
        for _ in 0..50 {
            let t = new_token(TOKEN_LENGTH);
            assert_eq!(t.len(), TOKEN_LENGTH);
            assert!(seen.insert(t));
        }
        assert!(!lan_ip().is_empty());
    }

    #[test]
    fn key_in_url() {
        let (server, _rx) = Server::start(0, Some("Correct Horse [battery] staple"), Vec::new()).unwrap();
        assert!(server.url().ends_with("/?k=Correct+Horse+%5Bbattery%5D+staple"), "{}", server.url());
        let (server, _rx) = Server::start(0, Some(""), Vec::new()).unwrap();
        assert_eq!(server.url(), server.address());
    }
}
